hack/generator/pkg/codegen/pipeline_create_storage_types.go

```diff
@@ -0,0 +1,212 @@
+	}
+
+	return visitedName, nil
+}
+
+func (factory *StorageTypeFactory) visitObjectType(
+	visitor *astmodel.TypeVisitor,
+	object *astmodel.ObjectType,
+	ctx interface{},
+) (astmodel.Type, error) {
+	vc := ctx.(StorageTypesVisitorContext)
+	oc := vc.forObject(object)
+
+	var errs []error
+	properties := object.Properties()
+	for i, prop := range properties {
+		propertyType, err := visitor.Visit(prop.PropertyType(), oc.forProperty(prop))
+		if err != nil {
+			errs = append(errs, err)
+		} else {
+			p := factory.makeStorageProperty(prop, propertyType)
+			properties[i] = p
+		}
+	}
+
+	if len(errs) > 0 {
+		err := kerrors.NewAggregate(errs)
+		return nil, err
+	}
+
+	ot := astmodel.NewObjectType().WithProperties(properties...)
+	return astmodel.MakeStorageType(ot), nil
+}
+
+func (factory *StorageTypeFactory) makeStorageProperty(prop *astmodel.PropertyDefinition, propertyType astmodel.Type) *astmodel.PropertyDefinition {
+	p := prop.WithType(propertyType).
+		MakeOptional().
+		WithoutValidation().
+		WithDescription("")
+
+	// If the property is *string we can simplify it to string
+	// This makes other generated code simpler
+	if ot, ok := p.PropertyType().(*astmodel.OptionalType); ok {
+		if ot.Element().Equals(astmodel.StringType) {
+			p = p.WithType(astmodel.StringType)
+		}
+	}
+
+	return p
+}
+
+// mapTypeName maps an existing type name into the right package for the matching storage type
+// Returns the original instance for any type that does not need to be mapped to storage
+func (factory *StorageTypeFactory) mapTypeName(name astmodel.TypeName) (*astmodel.TypeName, error) {
+	if !name.PackageReference.IsLocalPackage() {
+		// Don't need to map non-local packages
+		return &name, nil
+	}
+
+	storagePackage, err := astmodel.CreateStoragePackageReference(name.PackageReference)
+	if err != nil {
+		return nil, err
+	}
+
+	newName := astmodel.MakeTypeName(storagePackage, name.Name())
+	return &newName, nil
+}
+
+func (factory *StorageTypeFactory) visitArmType(
+	_ *astmodel.TypeVisitor,
+	it *astmodel.ArmType,
+	_ interface{},
+) (astmodel.Type, error) {
+	// We don't want to do anything with ARM types
+	return it, nil
+}
+
+func descriptionForStorageVariant(definition astmodel.TypeDefinition) []string {
+	_, pkg, err := definition.Name().PackageReference.GroupAndPackage()
+	if err != nil {
+		panic(err)
+	}
+
+	result := []string{
+		fmt.Sprintf("Storage version of %v.%v", pkg, definition.Name().Name()),
+	}
+	result = append(result, definition.Description()...)
+
+	return result
+}
+
+type StorageTypesVisitorContext struct {
+	object   *astmodel.ObjectType
+	property *astmodel.PropertyDefinition
+}
+
+func makeStorageTypesVisitorContext() StorageTypesVisitorContext {
+	return StorageTypesVisitorContext{}
+}
+
+func (context StorageTypesVisitorContext) forObject(object *astmodel.ObjectType) StorageTypesVisitorContext {
+	context.object = object
+	context.property = nil
```

Comment:
surprised this sets to `nil`?

Just wondering what context it wasn't `nil` and needs to be?

Reply:
Is it expected that this context only sets object + property once? If so, could panic if that were violated.

Reply:
Same with `forProperty`

Reply:
It creates a clone (note the receiver doesn't use a pointer) and then modifies it; if used for a different object, any property value would be stale - so I defensively set it to ensure nothing goes wrong.

Reply:
In _current use_ it's likely already nil, but it doesn't cost to set it and it prevents bugs if our usage pattern changes.

Reply:
I like this.